prov/tcp/src/tcpx_progress.c

```diff
@@ -29,17 +29,112 @@
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */
+
 #include <rdma/fi_errno.h>
 
 #include <prov.h>
 #include "tcpx.h"
 
 int tcpx_progress_close(struct tcpx_domain *domain)
+{
+	struct tcpx_progress *progress = domain->progress;
+
+
+	progress->do_progress = 0;
+	if (progress->progress_thread &&
+	    pthread_join(progress->progress_thread, NULL)) {
+		FI_WARN(&tcpx_prov, FI_LOG_DOMAIN,
+			"progress thread failed to join\n");
+	}
+	fd_signal_free(&progress->signal);
+	fastlock_destroy(&progress->signal_lock);
+	fi_epoll_close(progress->epoll_set);
+	return FI_SUCCESS;
+}
+
+void tcpx_progress_signal(struct tcpx_progress *progress)
+{
+	fastlock_acquire(&progress->signal_lock);
+	fd_signal_set(&progress->signal);
+	fastlock_release(&progress->signal_lock);
+}
+
+static int tcpx_progress_wait(struct tcpx_progress *progress)
+{
+	int ret;
+	void *ep_context;
+
+	ret = fi_epoll_wait(progress->epoll_set,
+			    &ep_context,1,-1);
+	if (ret < 0) {
+		FI_WARN(&tcpx_prov, FI_LOG_DOMAIN,
+			"Poll failed\n");
+		return -errno;
+	}
+
+	fastlock_acquire(&progress->signal_lock);
+	fd_signal_reset(&progress->signal);
+	fastlock_release(&progress->signal_lock);
+
+	return FI_SUCCESS;
+}
+
+static int tcpx_progress_wait_ok(struct tcpx_progress *progress)
 {
 	return -FI_ENOSYS;
 }
+
+void *tcpx_progress_thread(void *data)
+{
+	struct tcpx_progress *progress;
+	progress = (struct tcpx_progress *) data;
+	while (progress->do_progress) {
+		if (tcpx_progress_wait_ok(progress)) {
+			tcpx_progress_wait(progress);
+		}
+	}
+	return NULL;
+}
+
 int tcpx_progress_init(struct tcpx_domain *domain,
 		       struct tcpx_progress *progress)
 {
-	return -FI_ENOSYS;
+	int ret;
+	if (!progress)
+		return -FI_EINVAL;
```

Comment:
keep one blank line between variable declarations and code.  No need to check for null progress here.  That's a coding error not a runtime issue.